Backend/audio/xwax.1
.TH XWAX "1"
.SH NAME
xwax \- Digital vinyl on Linux
.SH SYNOPSIS
.B xwax
[\fIoptions\fR]
.SH DESCRIPTION
.P
xwax is vinyl emulation software for Linux. It allows DJs and
turntablists to playback digital audio files (MP3, Ogg Vorbis, FLAC,
AAC and more), controlled using a normal pair of turntables via
timecoded vinyls.
.SH OPTIONS
.P
The ordering of options is important. Most options apply to
subsequent music libraries or decks, which can be given multiple times.
See the
.B EXAMPLES
below.
.TP
.B \-l \fIpath\fR
Scan the music library or playlist at the given path.
.TP
.B \-t \fIname\fR
Use the named timecode for subsequent decks. See \-h for a list of
valid timecodes. You will need the corresponding timecode signal on
vinyl to control playback.
.TP
.B \-33
Set the reference playback speed for subsequent decks to 33 and one
third revolutions per minute. This is the default.
.TP
.B \-45
Set the reference playback speed for subsequent decks to 45
revolutions per minute.
.TP
.B \-c
Protect subsequent decks against certain operations during
playback.
.TP
.B \-u
Allow all operations on a deck during playback. This is the inverse
of the
.B \-c
option, and is the default.
.TP
.B \-\-phono
Adjust the noise thresholds of subsequent decks to tolerate a
cartridge-level signal connected to a line-level audio interface. This
is a 'software pre-amp'. Unless your audio path has low noise, this
will give worse results or may not work at all; a true phono
pre-amplifier is always preferred.
.TP
.B \-\-line
Set noise thresholds of subsequent decks to standard audio levels.
This reverses the effect of the
.B \-\-phono
option, and is the default.
.TP
.B \-i \fIpath\fR
Use the given importer executable for subsequent decks.
.TP
.B \-s \fIpath\fR
Use the given scanner executable to scan subsequent music libraries.
.TP
.B \-\-dummy
Create a deck which is not connected to any audio device, used
for testing.
.TP
.B \-k
Lock into RAM any memory required for real-time use.
This includes audio tracks held in memory which can be large.
Use
.B ulimit \-l
to raise the kernel's memory limit to allow this.
.TP
.B \-q \fIn\fR
Change the real-time priority of the process. A priority of 0 gives
the process no priority, and is used for testing only.
.TP
.B \-g [\fIn\fRx\fIn\fR][+\fIn\fR+\fIn\fR][/\fIf\fR]
Change the geometry of the display in size, position and scale (zoom)
respectively.
The size and position is passed
to SDL, which may use it to set the display mode, or size of an X window.
See the
.B EXAMPLES.
.TP
.B \-\-no\-decor
Request to the window manager to create a 'frameless' window which
does not have the regular controls such as title bars and buttons.
This can be useful in conjunction with the
.B \-g
flag for dedicated xwax installations.
.TP
.B \-h
Display the help message and default values.
.SH "ALSA DEVICE OPTIONS"
.P
The following options are available only when xwax is compiled with
ALSA support.
.TP
.B \-a \fIdevice\fR
Create a deck which uses the given ALSA device (eg. plughw:0).
.TP
.B \-r \fIhz\fR
Set the sample rate for subsequent decks.
.TP
.B \-m \fImilliseconds\fR
Set the ALSA buffer time for subsequent decks.
.SH "JACK DEVICE OPTIONS"
.P
The following options are available only when xwax is compiled with
JACK support.
.TP
.B \-j \fIname\fR
Create a deck which connects to JACK and registers under the given
name.
.P
xwax does not set the sample rate for JACK devices; it uses the sample
rate given in the global JACK configuration.
.SH "OSS DEVICE OPTIONS"
.P
The following options are available only when xwax is compiled with
OSS support.
.TP
.B \-d \fIpathname\fR
Create a deck which uses the given OSS device (eg. /dev/dsp).
.TP
.B \-r \fIhz\fR
Set the sample rate for subsequent decks.
.TP
.B \-b \fIn\fR
Set the number of OSS buffers for subsequent decks.
.TP
.B \-f \fIn\fR
Set the OSS buffer size (2^n bytes).
.SH HARDWARE CONTROLLER OPTIONS
.P
The following options are available only when xwax is compiled
with ALSA support.
.TP
.B \-\-dicer \fIdevice\fR
Use one or two Dicer controllers connected as the given ALSA device
(eg. hw:Dicer). See the section
.B NOVATION DICER CONTROLS
for more information.
.P
Adding a hardware controller results in control over subsequent decks,
up to the limit of the hardware.
.SH KEYBOARD CONTROLS
.P
The playback of each deck (direction, speed and position) is
controlled via the incoming timecode signal from the turntables.
The keyboard provides additional controls.
.P
"C-" and "S-" means a keypress is combined with
the 'Control' or 'Shift' key, respectively.
.P
Record selection controls:
.TP
cursor up, cursor down
Move highlighted record up/down by one.
.TP
page up, page down
Scroll the record listing up/down by one page.
.TP
left cursor, right cursor
Switch to the previous/next crate of records.
.TP
tab
Toggle between the current crate and the 'All records' crate.
.TP
C-tab
Toggle sort mode between: artist/track name, BPM and 'playlist'
order. Playlist order is the order in which records were returned
from the scanner.
.TP
C-S-tab
Re-scan the currently selected crate.
.P
To filter the current list of records type a portion of a record
name. Separate multiple searches with a space, and use backspace to
delete.
.P
Deck-specific controls:
.TS
l l l l.
Deck 0	Deck 1	Deck 2
F1	F5	F9	Load currently selected track to this deck
F2	F6	F10	Reset start of track to the current position
F3	F7	F11	Toggle timecode control on/off
C-F3	C-F7	C-F11	Cycle between available timecodes
.TE
.P
The "available timecodes" are those which have been the subject of any
.B \-t
flag on the command line.
Audio display controls:
.TP
+, \-
Zoom in/out the close-up audio meters for all decks.
.SH NOVATION DICER CONTROLS
.P
The Novation Dicer provides hardware control of cue points. The controls
are:
.TP
cue mode: dice button (1-5)
Jump to the specified cue point, or set it if unset.
.TP
loop-roll mode: dicer button (1-5)
"Punch" to the specified cue point, or set it if unset. Returns playback
to normal when the button is released.
.TP
mode button + dice button (1-5)
Clear the specified cue point.
.P
The dice buttons are lit to show that the corresponding cue point is
set.
.SH EXAMPLES
.P
2-deck setup using one directory of music and OSS devices:
.sp
.RS
xwax \-l ~/music \-d /dev/dsp \-d /dev/dsp1
.RE
.P
As above, but using ALSA devices:
.sp
.RS
xwax \-l ~/music \-d hw:0 \-d hw:1
.RE
.P
2-deck setup using a different timecode on each deck:
.sp
.RS
xwax \-l ~/music \-t serato_2a \-d hw:0 \-t mixvibes_v2 \-d hw:1
.RE
.P
As above, but with the second deck at 45 RPM:
.sp
.RS
xwax \-l ~/music \-t serato_2a \-d hw:0 \-t mixvibes_v2 \-45 \-d hw:1
.RE
.P
Default to the same timecode, but allow switching at runtime:
.sp
.RS
xwax \-l ~/music \-t serato_2a \-t mixvibes_v2 \-d hw:0 \-d hw:1
.RE
.P
3-deck setup with the third deck at a higher sample rate:
.sp
.RS
xwax \-l ~/music \-r 48000 \-a hw:0 \-a hw:1 \-r 96000 \-a hw:2
.RE
.P
Using all three device types simultaneously, one deck on each:
.sp
.RS
xwax \-l ~/music \-a hw:0 \-d /dev/dsp1 \-j jack0
.RE
.P
Scan multiple music libraries:
.sp
.RS
xwax \-l ~/music \-l ~/sounds \-l ~/mixes \-a hw:0
.RE
.P
Scan a second music library using a custom script:
.sp
.RS
xwax \-l ~/music \-i ./custom-scan \-l ~/sounds \-a hw:0
.RE
.P
Control two decks with Dicer hardware:
.sp
.RS
xwax \-\-dicer hw:Dicer \-a hw:0 \-a hw:1
.RE
.P
Use a high resolution and enlarge the user interface:
.sp
.RS
xwax -g 1920x1200/1.8 -a hw:0
.RE
.SH HOMEPAGE
http://xwax.org/
.SH AUTHOR
Mark Hills <mark@xwax.org>


Backend/audio/xwax.c
/*
 * Copyright (C) 2016 Mark Hills <mark@xwax.org>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * version 2, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 *
 * You should have received a copy of the GNU General Public License
 * version 2 along with this program; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 *
 */

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/mman.h> /* mlockall() */

#include <SDL.h> /* may override main() */

#include "alsa.h"
#include "controller.h"
#include "device.h"
#include "dicer.h"
#include "dummy.h"
#include "interface.h"
#include "jack.h"
#include "library.h"
#include "oss.h"
#include "realtime.h"
#include "thread.h"
#include "rig.h"
#include "timecoder.h"
#include "track.h"
#include "xwax.h"

#define DEFAULT_OSS_BUFFERS 8
#define DEFAULT_OSS_FRAGMENT 7

#define DEFAULT_ALSA_BUFFER 128 /* milliseconds */

#define DEFAULT_RATE 44100
#define DEFAULT_PRIORITY 80

#define DEFAULT_IMPORTER EXECDIR "/xwax-import"
#define DEFAULT_SCANNER EXECDIR "/xwax-scan"
#define DEFAULT_TIMECODE "serato_2a"

#define ARRAY_SIZE(x) (sizeof(x) / sizeof(*x))

char *banner = "xwax " VERSION \
    " (C) Copyright 2016 Mark Hills <mark@xwax.org>";

size_t ndeck;
struct deck deck[3];

static size_t nctl;
static struct controller ctl[2];

static struct rt rt;

static double speed;
static bool protect, phono;
static const char *importer;
static struct timecode_def *timecode;

static char *single_track;

static void usage(FILE *fd)
{
    fprintf(fd, "Usage: xwax [<options>]\n\n");

    fprintf(fd, "Program-wide options:\n"
      "  -k             Lock real-time memory into RAM\n"
      "  -q <n>         Real-time priority (0 for no priority, default %d)\n"
      "  -g <s>         Set display geometry (see man page)\n"
      "  --no-decor     Request a window with no decorations\n"
      "  -h             Display this message to stdout and exit\n\n",
      DEFAULT_PRIORITY);

    fprintf(fd, "Music library options:\n"
      "  -p <path>      Path to a song\n"
      "  -l <path>      Location to scan for audio tracks\n"
      "  -s <program>   Library scanner (default '%s')\n\n",
      DEFAULT_SCANNER);

    fprintf(fd, "Deck options:\n"
      "  -t <name>      Timecode name\n"
      "  -33            Use timecode at 33.3RPM (default)\n"
      "  -45            Use timecode at 45RPM\n"
      "  -c             Protect against certain operations while playing\n"
      "  -u             Allow all operations when playing\n"
      "  --line         Line level signal (default)\n"
      "  --phono        Tolerate cartridge level signal ('software pre-amp')\n"
      "  -i <program>   Importer (default '%s')\n"
      "  --dummy        Build a dummy deck with no audio device\n\n",
      DEFAULT_IMPORTER);

#ifdef WITH_OSS
    fprintf(fd, "OSS device options:\n"
      "  -d <device>    Build a deck connected to OSS audio device\n"
      "  -r <hz>        Sample rate (default %dHz)\n"
      "  -b <n>         Number of buffers (default %d)\n"
      "  -f <n>         Buffer size to request (2^n bytes, default %d)\n\n",
      DEFAULT_RATE, DEFAULT_OSS_BUFFERS, DEFAULT_OSS_FRAGMENT);
#endif

#ifdef WITH_ALSA
    fprintf(fd, "ALSA device options:\n"
      "  -a <device>    Build a deck connected to ALSA audio device\n"
      "  -r <hz>        Sample rate (default %dHz)\n"
      "  -m <ms>        Buffer time (default %dms)\n\n",
      DEFAULT_RATE, DEFAULT_ALSA_BUFFER);
#endif

#ifdef WITH_JACK
    fprintf(fd, "JACK device options:\n"
      "  -j <name>      Create a JACK deck with the given name\n\n");
#endif

#ifdef WITH_ALSA
    fprintf(fd, "MIDI control:\n"
      "  --dicer <dev>   Novation Dicer\n\n");
#endif

    fprintf(fd,
      "The ordering of options is important. Options apply to subsequent\n"
      "music libraries or decks, which can be given multiple times. See the\n"
      "manual for details.\n\n"
      "Available timecodes (for use with -t):\n"
      "  serato_2a (default), serato_2b, serato_cd,\n"
      "  traktor_a, traktor_b, mixvibes_v2, mixvibes_7inch\n\n"
      "See the xwax(1) man page for full information and examples.\n");
}

static struct device* start_deck(const char *desc)
{
    fprintf(stderr, "Initialising deck %zd (%s)...\n", ndeck, desc);

    if (ndeck == ARRAY_SIZE(deck)) {
        fprintf(stderr, "Too many decks.\n");
        return NULL;
    }

    return &deck[ndeck].device;
}

static int commit_deck(void)
{
    int r;
    struct deck *d;
    size_t n;

    /* Fallback to a default timecode. Don't initialise this at the
     * front of the program to avoid buildling unnecessary LUTs */

    if (timecode == NULL) {
        timecode = timecoder_find_definition(DEFAULT_TIMECODE);
        assert(timecode != NULL);
    }

    d = &deck[ndeck];

    r = deck_init(d, &rt, timecode, importer, speed, phono, protect);
    if (r == -1)
        return -1;

    /* Connect this deck to available controllers */

    for (n = 0; n < nctl; n++)
        controller_add_deck(&ctl[n], d);

    ndeck++;

    return 0;
}

int main(int argc, char *argv[])
{
    int rc = -1, n, priority;
    const char *scanner, *geo;
    char *endptr;
    bool use_mlock, decor;

    struct library library;

#if defined WITH_OSS || WITH_ALSA
    int rate;
#endif

#ifdef WITH_OSS
    int oss_buffers, oss_fragment;
#endif

#ifdef WITH_ALSA
    int alsa_buffer;
#endif

    fprintf(stderr, "%s\n\n" NOTICE "\n\n", banner);

    if (thread_global_init() == -1)
        return -1;

    if (rig_init() == -1)
        return -1;
    rt_init(&rt);
    library_init(&library);

    ndeck = 0;
    geo = "";
    decor = true;
    nctl = 0;
    priority = DEFAULT_PRIORITY;
    importer = DEFAULT_IMPORTER;
    scanner = DEFAULT_SCANNER;
    timecode = NULL;
    speed = 1.0;
    protect = false;
    phono = false;
    use_mlock = false;

#if defined WITH_OSS || WITH_ALSA
    rate = DEFAULT_RATE;
#endif

#ifdef WITH_ALSA
    alsa_buffer = DEFAULT_ALSA_BUFFER;
#endif

#ifdef WITH_OSS
    oss_fragment = DEFAULT_OSS_FRAGMENT;
    oss_buffers = DEFAULT_OSS_BUFFERS;
#endif

    /* Skip over command name */

    argv++;
    argc--;

    while (argc > 0) {

        if (!strcmp(argv[0], "-h")) {
            usage(stdout);
            return 0;

#ifdef WITH_OSS
        } else if (!strcmp(argv[0], "-f")) {

            /* Set fragment size for subsequent devices */

            if (argc < 2) {
                fprintf(stderr, "-f requires an integer argument.\n");
                return -1;
            }

            oss_fragment = strtol(argv[1], &endptr, 10);
            if (*endptr != '\0') {
                fprintf(stderr, "-f requires an integer argument.\n");
                return -1;
            }

            /* Fragment sizes greater than the default aren't useful
             * as they are dependent on DEVICE_FRAME */

            if (oss_fragment < DEFAULT_OSS_FRAGMENT) {
                fprintf(stderr, "Fragment size must be %d or more; aborting.\n",
                        DEFAULT_OSS_FRAGMENT);
                return -1;
            }

            argv += 2;
            argc -= 2;

        } else if (!strcmp(argv[0], "-b")) {

            /* Set number of buffers for subsequent devices */

            if (argc < 2) {
                fprintf(stderr, "-b requires an integer argument.\n");
                return -1;
            }

            oss_buffers = strtol(argv[1], &endptr, 10);
            if (*endptr != '\0') {
                fprintf(stderr, "-b requires an integer argument.\n");
                return -1;
            }

            argv += 2;
            argc -= 2;
#endif

#if defined WITH_OSS || WITH_ALSA
        } else if (!strcmp(argv[0], "-r")) {

            /* Set sample rate for subsequence devices */

            if (argc < 2) {
                fprintf(stderr, "-r requires an integer argument.\n");
                return -1;
            }

            rate = strtol(argv[1], &endptr, 10);
            if (*endptr != '\0') {
                fprintf(stderr, "-r requires an integer argument.\n");
                return -1;
            }

            argv += 2;
            argc -= 2;
#endif

#ifdef WITH_ALSA
        } else if (!strcmp(argv[0], "-m")) {

            /* Set size of ALSA buffer for subsequence devices */

            if (argc < 2) {
                fprintf(stderr, "-m requires an integer argument.\n");
                return -1;
            }

            alsa_buffer = strtol(argv[1], &endptr, 10);
            if (*endptr != '\0') {
                fprintf(stderr, "-m requires an integer argument.\n");
                return -1;
            }

            argv += 2;
            argc -= 2;
#endif

        } else if (!strcmp(argv[0], "-d") || !strcmp(argv[0], "-a") ||
		  !strcmp(argv[0], "-j"))
	{
            int r;
            struct device *device;

            /* Create a deck */

            if (argc < 2) {
                fprintf(stderr, "-%c requires a device name as an argument.\n",
                        argv[0][1]);
                return -1;
            }

            device = start_deck(argv[1]);
            if (device == NULL)
                return -1;

            /* Work out which device type we are using, and initialise
             * an appropriate device. */

            switch(argv[0][1]) {

#ifdef WITH_OSS
            case 'd':
                r = oss_init(device, argv[1], rate, oss_buffers, oss_fragment);
                break;
#endif
#ifdef WITH_ALSA
            case 'a':
                r = alsa_init(device, argv[1], rate, alsa_buffer);
                break;
#endif
#ifdef WITH_JACK
            case 'j':
                r = jack_init(device, argv[1]);
                break;
#endif
            default:
                fprintf(stderr, "Device type is not supported by this "
                        "distribution of xwax.\n");
                return -1;
            }

            if (r == -1)
                return -1;

            commit_deck();

            argv += 2;
            argc -= 2;

        } else if (!strcmp(argv[0], "--dummy")) {

            struct device *v;

            v = start_deck("dummy");
            if (v == NULL)
                return -1;

            dummy_init(v);
            commit_deck();

            argv++;
            argc--;

        } else if (!strcmp(argv[0], "-t")) {

            /* Set the timecode definition to use */

            if (argc < 2) {
                fprintf(stderr, "-t requires a name as an argument.\n");
                return -1;
            }

            timecode = timecoder_find_definition(argv[1]);
            if (timecode == NULL) {
                fprintf(stderr, "Timecode '%s' is not known.\n", argv[1]);
                return -1;
            }

            argv += 2;
            argc -= 2;

        } else if (!strcmp(argv[0], "-p")) {

            /* set the track to use */

            if (argc < 2) {
                fprintf(stderr, "-p requires a name as an argument.\n");
                return -1;
            }

            single_track = argv[1];

            argv += 2;
            argc -= 2;

        } else if (!strcmp(argv[0], "-33")) {

            speed = 1.0;

            argv++;
            argc--;

        } else if (!strcmp(argv[0], "-45")) {

            speed = 1.35;

            argv++;
            argc--;

        } else if (!strcmp(argv[0], "-c")) {

            protect = true;

            argv++;
            argc--;

        } else if (!strcmp(argv[0], "-u")) {

            protect = false;

            argv++;
            argc--;

        } else if (!strcmp(argv[0], "--line")) {

            phono = false;

            argv++;
            argc--;

        } else if (!strcmp(argv[0], "--phono")) {

            phono = true;

            argv++;
            argc--;

        } else if (!strcmp(argv[0], "-k")) {

            use_mlock = true;
            track_use_mlock();

            argv++;
            argc--;

        } else if (!strcmp(argv[0], "-q")) {

            if (argc < 2) {
                fprintf(stderr, "-q requires an integer argument.\n");
                return -1;
            }

            priority = strtol(argv[1], &endptr, 10);
            if (*endptr != '\0') {
                fprintf(stderr, "-q requires an integer argument.\n");
                return -1;
            }

            if (priority < 0) {
                fprintf(stderr, "Priority (%d) must be zero or positive.\n",
                        priority);
                return -1;
            }

            argv += 2;
            argc -= 2;

        } else if (!strcmp(argv[0], "-g")) {

            if (argc < 2) {
                fprintf(stderr, "-g requires an argument.\n");
                return -1;
            }

            geo = argv[1];

            argv += 2;
            argc -= 2;

        } else if (!strcmp(argv[0], "--no-decor")) {

            decor = false;

            argv++;
            argc--;

        } else if (!strcmp(argv[0], "-i")) {

            /* Importer script for subsequent decks */

            if (argc < 2) {
                fprintf(stderr, "-i requires an executable path "
                        "as an argument.\n");
                return -1;
            }

            importer = argv[1];

            argv += 2;
            argc -= 2;

        } else if (!strcmp(argv[0], "-s")) {

            /* Scan script for subsequent libraries */

            if (argc < 2) {
                fprintf(stderr, "-s requires an executable path "
                        "as an argument.\n");
                return -1;
            }

            scanner = argv[1];

            argv += 2;
            argc -= 2;

        } else if (!strcmp(argv[0], "-l")) {

            /* Load in a music library */

            if (argc < 2) {
                fprintf(stderr, "-l requires a pathname as an argument.\n");
                return -1;
            }

            if (library_import(&library, scanner, argv[1]) == -1)
                return -1;

            argv += 2;
            argc -= 2;

#ifdef WITH_ALSA
        } else if (!strcmp(argv[0], "--dicer")) {

            struct controller *c;

            if (nctl == sizeof ctl) {
                fprintf(stderr, "Too many controllers; aborting.\n");
                return -1;
            }

            c = &ctl[nctl];

            if (argc < 2) {
                fprintf(stderr, "Dicer requires an ALSA device name.\n");
                return -1;
            }

            if (dicer_init(c, &rt, argv[1]) == -1)
                return -1;

            nctl++;

            argv += 2;
            argc -= 2;
#endif

        } else {
            fprintf(stderr, "'%s' argument is unknown; try -h.\n", argv[0]);
            return -1;
        }
    }

#ifdef WITH_ALSA
    alsa_clear_config_cache();
#endif

    if (ndeck == 0) {
        fprintf(stderr, "You need to give at least one audio device to use "
                "as a deck; try -h.\n");
        return -1;
    }

    rc = EXIT_FAILURE; /* until clean exit */

    /* Order is important: launch realtime thread first, then mlock.
     * Don't mlock the interface, use sparingly for audio threads */

    if (rt_start(&rt, priority) == -1)
        return -1;

    if (use_mlock && mlockall(MCL_CURRENT) == -1) {
        perror("mlockall");
        goto out_rt;
    }

//    if (interface_start(&library, geo, decor) == -1)
//        goto out_rt;


    struct record re;
    re.pathname = single_track;
    deck_load(&deck[0], &re);
    deck_recue(&deck[0]);

    if (rig_main() == -1)
        goto out_interface;

    rc = EXIT_SUCCESS;
    fprintf(stderr, "Exiting cleanly...\n");

out_interface:
    interface_stop();
out_rt:
    rt_stop(&rt);

    for (n = 0; n < ndeck; n++)
        deck_clear(&deck[n]);

    for (n = 0; n < nctl; n++)
        controller_clear(&ctl[n]);

    timecoder_free_lookup();
    library_clear(&library);
    rt_clear(&rt);
    rig_clear();
    thread_global_clear();

    if (rc == EXIT_SUCCESS)
        fprintf(stderr, "Done.\n");

    return rc;
}
